api/src/main/java/net/kyori/adventure/text/JoinConfigurationImpl.java

```diff
@@ -0,0 +1,234 @@
+    this.lastSeparator = builder.lastSeparator == null ? null : builder.lastSeparator.asComponent();
+    this.prefix = builder.prefix == null ? null : builder.prefix.asComponent();
+    this.suffix = builder.suffix == null ? null : builder.suffix.asComponent();
+    this.operator = builder.operator;
+    this.lastSeparatorIfSerial = builder.lastSeparatorIfSerial == null ? null : builder.lastSeparatorIfSerial.asComponent();
+  }
+
+  @Override
+  public @Nullable Component prefix() {
+    return this.prefix;
+  }
+
+  @Override
+  public @Nullable Component suffix() {
+    return this.suffix;
+  }
+
+  @Override
+  public @Nullable Component separator() {
+    return this.separator;
+  }
+
+  @Override
+  public @Nullable Component lastSeparator() {
+    return this.lastSeparator;
+  }
+
+  @Override
+  public @NotNull UnaryOperator<Component> operator() {
+    return this.operator;
+  }
+
+  @Override
+  public @Nullable Component lastSeparatorIfSerial() {
+    return this.lastSeparatorIfSerial;
+  }
+
+  @Override
+  public JoinConfiguration.@NotNull Builder toBuilder() {
+    return new BuilderImpl(this);
+  }
+
+  @Override
+  public @NotNull Stream<? extends ExaminableProperty> examinableProperties() {
+    return Stream.of(
+      ExaminableProperty.of("separator", this.separator),
+      ExaminableProperty.of("lastSeparator", this.lastSeparator),
+      ExaminableProperty.of("prefix", this.prefix),
+      ExaminableProperty.of("suffix", this.suffix),
+      ExaminableProperty.of("operator", this.operator)
+    );
+  }
+
+  @Override
+  public String toString() {
+    return this.examine(StringExaminer.simpleEscaping());
+  }
+
+  @Contract(pure = true)
+  static @NotNull Component join(final @NotNull JoinConfiguration config, final @NotNull Iterable<? extends ComponentLike> components) {
+    Objects.requireNonNull(config, "config");
+    Objects.requireNonNull(components, "components");
+
+    final Iterator<? extends ComponentLike> it = components.iterator();
+    final Component prefix = config.prefix();
+    final Component suffix = config.suffix();
+    final UnaryOperator<Component> operator = config.operator();
+
+    if (!it.hasNext()) {
+      if (prefix == null && suffix == null) return Component.empty();
+
+      final TextComponent.Builder builder = Component.text();
+      if (prefix != null) builder.append(prefix);
+      if (suffix != null) builder.append(suffix);
+      return builder.build();
+    }
+
+    ComponentLike component = it.next();
+    int componentsSeen = 0;
+
+    if (!it.hasNext() && prefix == null && suffix == null) return operator.apply(component.asComponent());
+
+    final Component separator = config.separator();
+    final boolean hasSeparator = separator != null;
+
+    final TextComponent.Builder builder = Component.text();
+    if (prefix != null) builder.append(prefix);
+
+    while (component != null) {
+      builder.append(operator.apply(component.asComponent()));
```

Comment:
Similarly, we should be asserting that this isn't null or documenting that we allow it and handling it.